codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -28,163 +23,157 @@
+                }
+            }
+            ReferenceCountUtil.release(msg);
+        } else {
+            throw new UnsupportedMessageTypeException(msg);
+        }
+    }
 
-        UpgradeEvent upgrade = (UpgradeEvent) evt;
-        ctx.fireUserEventTriggered(upgrade.retain());
-        try {
-            Http2Stream stream = http2Handler.connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
-            // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
-            // The stream was already made active, but ctx may have been null so it wasn't initialized.
-            // https://github.com/netty/netty/issues/4942
-            new ConnectionListener().onStreamActive(stream);
-            upgrade.upgradeRequest().headers().setInt(
-                HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
-            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
-                .channelRead(ctx, upgrade.upgradeRequest().retain());
-        } finally {
-            upgrade.release();
+    void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        childChannel.fireChildRead(frame);
+        if (!childChannel.inStreamsToFireChildReadComplete) {
+            channelsToFireChildReadComplete.add(childChannel);
+            childChannel.inStreamsToFireChildReadComplete = true;
         }
     }
 
-    // Override this to signal it will never throw an exception.
     @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        ctx.fireExceptionCaught(cause);
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof Http2StreamStateEvent)) {
+            ctx.fireUserEventTriggered(evt);
+            return;
+        }
+
+        int streamId = ((Http2StreamStateEvent) evt).streamId();
+        if (evt instanceof Http2StreamActiveEvent) {
+            onStreamActive(streamId);
+        } else if (evt instanceof Http2StreamClosedEvent) {
+            onStreamClosed(streamId);
+        } else {
+            throw new UnsupportedMessageTypeException(evt);
+        }
     }
 
-    // Override this to signal it will never throw an exception.
-    @Override
-    public void flush(ChannelHandlerContext ctx) {
-        ctx.flush();
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        Http2StreamChannel prevValue = childChannels.put(streamId, childChannel);
+        // TODO: Probably not assert but proper error handling.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)